File: dotnet/ThirtyFour.UI/Form/RadioButton.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using OpenQA.Selenium;

namespace ThirtyFour.UI.Form
{
    /// <summary>
    /// Represents a HTML radio button element
    /// </summary>
    public class RadioButton : BooleanElement
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="element">The wrapped element</param>
        public RadioButton(IWebElement element) : base(element) { }


        /// <summary>
        /// Checks the radio button
        /// </summary>
        public void SetChecked()
        {
            if (!IsChecked)
            {
                Element.Click();
            }
        }

        /// <summary>
        /// Returns all radio buttons in the group    
        /// </summary>
        /// <returns>A ReadOnlyCollection of all buttons in the group</returns>
        public IReadOnlyCollection<RadioButton> GetGroupButtons()
        {
            var buttons = new List<RadioButton>();
            var buttonName = Element.GetAttribute("name");
            if (buttonName == null)
            {
                buttons.Add(this);

            }
            else
            {
                var buttonElements = Driver.FindElements(By.CssSelector(String.Format("input[type='radio'][name='{0}']", buttonName)));
                foreach (IWebElement el in buttonElements)
                {
                    buttons.Add(new RadioButton(el));
                }
            }
            return new ReadOnlyCollection<RadioButton>(buttons);
        }

        /// <summary>
        /// Gets the selected radio button from the the group
        /// </summary>
        /// <returns>the selected radio button</returns>
        public RadioButton GetSelectedGroupButton()
        {
            var buttons = GetGroupButtons();
            foreach (RadioButton button in buttons)
            {
                if (button.IsChecked)
                {
                    return button;
                }
            }
            return null;
        }

        /// <summary>
        /// Gets the value of the selected radio button in the same group
        /// </summary>
        /// <returns>The value of the selected radio button or an empty string</returns>
        public string GetGroupValue()
        {
            var button = GetSelectedGroupButton();
            if (button == null)
            {
                return "";
            }
            return button.Value;
        }

        /// <summary>
        /// Checks a radio button in the same group which has a matching value
        /// </summary>
        /// <param name="value">the value fo the radio button</param>
        public void SetCheckedByValue(string value)
        {
            var buttons = GetGroupButtons();
            foreach (RadioButton button in buttons)
            {
                if (button.Value.Equals(value))
                {
                    button.SetChecked();
                    return;
                }
            }
            throw new NoSuchElementException("Unable to find radio button with value " + value);
        }

        /// <summary>
        /// Checks a radio button in the same group which has a matching label
        /// </summary>
        /// <param name="label">the value fo the radio button</param>
        public void SetCheckedByLabel(string label)
        {
            var buttons = GetGroupButtons();
            foreach (RadioButton button in buttons)
            {
                if (button.LabelText.Equals(label))
                {
                    button.SetChecked();
                    return;
                }
            }
            throw new NoSuchElementException("Unable to find radio button with label " + label);
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/form/RadioButton.java
package com.github.danhumphrey.thirtyfour.ui.form;

import java.util.ArrayList;
import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;

/**
 * The RadioBUtton class provides helper methods for interacting with a HTML radio button and other radio buttons in the same group
 */
public class RadioButton extends BooleanElement {

	/**
	 * Constructor
	 * @param element the wrapped radio button element 
	 */
	public RadioButton(WebElement element) {
		super(element);
	}
	
	/**
	 * Checks the radio button
	 */
	public void setChecked() {
		if(!this.element.isSelected()) {
			this.element.click();
		}
	}

	/**
	 * Gets a list of all radio buttons in the group
	 * @return a list of all buttons in the group
	 */
	public List<RadioButton> getGroupButtons() {
		List<RadioButton> buttons = new ArrayList<RadioButton>();
		String buttonName = this.element.getAttribute("name");
		if(buttonName == null) {
			buttons.add(this);
		} else {
			List<WebElement> allElements = this.driver.findElements(By.cssSelector((String.format("input[type='radio'][name='%s']",buttonName))));
			for(WebElement el : allElements) {
				buttons.add(new RadioButton(el));
			}
		}
		return buttons;
	}
	
	/**
	 * Gets the selected radio button from the the group
	 * @return the selected radio button or null none are selected
	 */
	public RadioButton getSelectedGroupButton() {
		List<RadioButton> buttons = this.getGroupButtons();
		for(RadioButton button :  buttons) {
			if(button.getChecked()) {
				return button;
			}
		}
		return null;
	}
	
	/**
	 * Gets the value of the selected radio button in the group
	 * @return the value of the selected radio button or an empty string
	 */
	public String getGroupValue() {
		
		RadioButton button = this.getSelectedGroupButton();
		if(button == null) {
			return "";
		}
		return button.getValue();
	}

	/**
	 * Checks a radio button in the same group which has a matching value
	 * @param value the value of the radio button
	 * @throws NoSuchElementException if a matching radio button cannot be found
	 */
	public void setCheckedByValue(String value) throws NoSuchElementException{
		List<RadioButton> buttons = this.getGroupButtons();
		for(RadioButton button :  buttons) {
			if(button.getValue().equals(value)) {
				button.setChecked();
				return;
			}
		}
		throw new NoSuchElementException("Unable to find radio button with value " + value);
	}

	/**
	 * Checks a radio button in the same group which has a matching value
	 * @param label the label of the radio button
	 * @throws NoSuchElementException if a matching radio button cannot be found
	 */
	public void setCheckedByLabel(String label) throws NoSuchElementException{
		List<RadioButton> buttons = this.getGroupButtons();
		for(RadioButton button :  buttons) {
			if(button.getLabelText().equals(label)) {
				button.setChecked();
				return;
			}
		}
		throw new NoSuchElementException("Unable to find Radio button with label " + label);
		
	}
}
